### PR TITLE
qt6, python3Packages.pyqt6: unbreak on darwin

### LINKED ISSUE #210905: qt6: broken on darwin
### Steps To Reproduce
1. build *qt6.qtbase*

### Build log
```
error: Package ‘mtdev-1.1.6’ in /Users/weijiawang/Projects/nixpkgs/pkgs/development/libraries/mtdev/default.nix:14 is not supported on ‘aarch64-darwin’, refusing to evaluate.
```

### Additional context

The regression is due to #192057, which forgot to maintain the `stdenv.isLinux` guard on darwin. While it is straightforward to fix the aarch64-darwin build, it isn't enough for x86_64-darwin due to #210861.

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

### PR DESCRIPTION
**Pull request by wegank**
###### Description of changes

The added qmake patch mimics [this](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/qt-5/5.15/qtbase.patch.d/0001-qtbase-mkspecs-mac.patch) in Qt 5.

Closes #210905.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by wegank** [comment]
@ofborg build python3Packages.pyqt6

File: pkgs/development/libraries/qt-6/modules/qtbase.nix [hunk 1]
```diff
@@ -222,8 +222,8 @@ stdenv.mkDerivation rec {
     "-DQT_FEATURE_journald=ON"
     "-DQT_FEATURE_vulkan=ON"
   ] ++ lib.optionals stdenv.isDarwin [
-    # build as a set of dynamic libraries
-    "-DFEATURE_framework=OFF"
```
Comment (NickCao) [inline pkgs/development/libraries/qt-6/modules/qtbase.nix:226]: What's the benefits/implications of building as frameworks?

File: pkgs/development/libraries/qt-6/modules/qtbase.nix [hunk 2]
```diff
@@ -269,8 +269,10 @@ stdenv.mkDerivation rec {
     moveToOutput libexec "$dev"
 
     # fixup .pc file (where to find 'moc' etc.)
-    sed -i "$dev/lib/pkgconfig/Qt6Core.pc" \
-      -e "/^bindir=/ c bindir=$dev/bin"
+    if [ -d "$dev/lib/pkgconfig" ]; then
+      sed -i "$dev/lib/pkgconfig/Qt6Core.pc" \
+        -e "/^bindir=/ c bindir=$dev/bin"
+    fi
```
Comment (NickCao) [inline pkgs/development/libraries/qt-6/modules/qtbase.nix:275]:
```suggestion
    if [ -f "$dev/lib/pkgconfig/Qt6Core.pc" ]; then
      sed -i "$dev/lib/pkgconfig/Qt6Core.pc" \
        -e "/^bindir=/ c bindir=$dev/bin"
    fi
```

File: pkgs/development/libraries/qt-6/modules/qtbase.nix (hunk 1 above)
Comment (wegank) [reply pkgs/development/libraries/qt-6/modules/qtbase.nix:226]: qmake generates erroneous link args (e.g. `-lQtNetwork.1`) when not building as frameworks on darwin, so I don't have much choice...

File: pkgs/development/libraries/qt-6/modules/qtbase.nix (hunk 2 above)
Comment (wegank) [reply pkgs/development/libraries/qt-6/modules/qtbase.nix]: Done. Feel free to squash the commits.

**Review by NickCao — APPROVED** [review]
You may have to do a rebase if https://github.com/NixOS/nixpkgs/pull/210134 is merged first.

**Comment by LunNova** [comment]
Result of `nixpkgs-review pr 204724` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>obs-studio-plugins.obs-hyperion</li>
  </ul>
</details>
<details>
  <summary>93 packages built:</summary>
  <ul>
    <li>adwaita-qt6</li>
    <li>beamerpresenter (beamerpresenter-mupdf)</li>
    <li>beamerpresenter-poppler</li>
    <li>calcmysky</li>
    <li>calibre</li>
    <li>cudaPackages.nsight_compute</li>
    <li>cutemaze</li>
    <li>duckstation</li>
    <li>fcitx5-chinese-addons</li>
    <li>fcitx5-configtool</li>
    <li>fcitx5-unikey</li>
    <li>fcitx5-with-addons</li>
    <li>gpxsee (gpxsee-qt5)</li>
    <li>gpxsee-qt6</li>
    <li>gqrx</li>
    <li>gqrx-gr-audio</li>
    <li>gqrx-portaudio</li>
    <li>jami-client</li>
    <li>ladybird</li>
    <li>libsForQt5.fcitx5-qt (plasma5Packages.fcitx5-qt)</li>
    <li>libsForQt5.qtpbfimageplugin (plasma5Packages.qtpbfimageplugin)</li>
    <li>maestral-gui</li>
    <li>mediaelch-qt6</li>
    <li>mozillavpn</li>
    <li>obs-studio</li>
    <li>obs-studio-plugins.droidcam-obs</li>
    <li>obs-studio-plugins.input-overlay</li>
    <li>obs-studio-plugins.looking-glass-obs</li>
    <li>obs-studio-plugins.obs-gstreamer</li>
    <li>obs-studio-plugins.obs-move-transition</li>
    <li>obs-studio-plugins.obs-multi-rtmp</li>
    <li>obs-studio-plugins.obs-nvfbc</li>
    <li>obs-studio-plugins.obs-pipewire-audio-capture</li>
    <li>obs-studio-plugins.obs-source-record</li>
    <li>obs-studio-plugins.obs-vkcapture</li>
    <li>obs-studio-plugins.wlrobs</li>
    <li>pokefinder</li>
    <li>prismlauncher</li>
    <li>python310Packages.pyqt6</li>
    <li>python310Packages.pyqt6-webengine</li>
    <li>python39Packages.pyqt6</li>
    <li>python39Packages.pyqt6-webengine</li>
    <li>qMasterPassword</li>
    <li>qalculate-qt</li>
    <li>qgnomeplatform-qt6</li>
    <li>qmmp</li>
    <li>qt6.full (qt6Packages.full)</li>
    <li>qt6.qmake (qt6Packages.qmake)</li>
    <li>qt6.qt3d (qt6Packages.qt3d)</li>
    <li>qt6.qt5compat (qt6Packages.qt5compat)</li>
    <li>qt6.qtbase (qt6Packages.qtbase)</li>
    <li>qt6.qtcharts (qt6Packages.qtcharts)</li>
    <li>qt6.qtconnectivity (qt6Packages.qtconnectivity)</li>
    <li>qt6.qtdatavis3d (qt6Packages.qtdatavis3d)</li>
    <li>qt6.qtdeclarative (qt6Packages.qtdeclarative)</li>
    <li>qt6.qtdoc (qt6Packages.qtdoc)</li>
    <li>qt6.qthttpserver (qt6Packages.qthttpserver)</li>
    <li>qt6.qtimageformats (qt6Packages.qtimageformats)</li>
    <li>qt6.qtlanguageserver (qt6Packages.qtlanguageserver)</li>
    <li>qt6.qtlottie (qt6Packages.qtlottie)</li>
    <li>qt6.qtmultimedia (qt6Packages.qtmultimedia)</li>
    <li>qt6.qtnetworkauth (qt6Packages.qtnetworkauth)</li>
    <li>qt6.qtpositioning (qt6Packages.qtpositioning)</li>
    <li>qt6.qtquick3d (qt6Packages.qtquick3d)</li>
    <li>qt6.qtquick3dphysics (qt6Packages.qtquick3dphysics)</li>
    <li>qt6.qtquicktimeline (qt6Packages.qtquicktimeline)</li>
    <li>qt6.qtremoteobjects (qt6Packages.qtremoteobjects)</li>
    <li>qt6.qtscxml (qt6Packages.qtscxml)</li>
    <li>qt6.qtsensors (qt6Packages.qtsensors)</li>
    <li>qt6.qtserialbus (qt6Packages.qtserialbus)</li>
    <li>qt6.qtserialport (qt6Packages.qtserialport)</li>
    <li>qt6.qtshadertools (qt6Packages.qtshadertools)</li>
    <li>qt6.qtspeech (qt6Packages.qtspeech)</li>
    <li>qt6.qtsvg (qt6Packages.qtsvg)</li>
    <li>qt6.qttools (qt6Packages.qttools)</li>
    <li>qt6.qttranslations (qt6Packages.qttranslations)</li>
    <li>qt6.qtvirtualkeyboard (qt6Packages.qtvirtualkeyboard)</li>
    <li>qt6.qtwayland (qt6Packages.qtwayland)</li>
    <li>qt6.qtwebchannel (qt6Packages.qtwebchannel)</li>
    <li>qt6.qtwebengine (qt6Packages.qtwebengine)</li>
    <li>qt6.qtwebsockets (qt6Packages.qtwebsockets)</li>
    <li>qt6.qtwebview (qt6Packages.qtwebview)</li>
    <li>qt6Packages.kcoreaddons</li>
    <li>qt6Packages.poppler</li>
    <li>qt6Packages.qtpbfimageplugin</li>
    <li>qt6Packages.quazip</li>
    <li>qt6Packages.qxlsx</li>
    <li>qutebrowser-qt6</li>
    <li>retext</li>
    <li>stellarium</li>
    <li>tdesktop</li>
    <li>zeal-qt6</li>
    <li>zint</li>
  </ul>
</details>

**Review by LunNova — APPROVED** [review]

**Comment by lilyinstarlight** [comment]
Ah, I hadn't seen this PR before I made #210134 and I targeted that to staging before I realized the qt6-only rebuilds aren't actually that many (and it has already been merged to staging)

Given this PR has conflicts with that one, but is targeted to a different branch, would it be easier to just retarget this to staging and fix the merge conflict, or is there a fancier method to fix the merge conflict in advance for the different branches?

**Comment by NickCao** [comment]
Ah I didn't see that was targeting staging. No need to worry about that, the conflict is minor and can be easily resolved later.

**Comment by wegank** [comment]
Oh no, there is a regression from #192057 that prevents building qtbase on darwin...

**Comment by wegank** [comment]
@ofborg build qt6.qtbase

**Comment by wegank** [comment]
No, qt6.qtbase is broken on master...

**Comment by wegank** [comment]
@orborg build python3Packages.pyqt6

**Comment by NickCao** [comment]
Apart from the libepoxy regression on `x86_64-darwin`, this looks good to go.

**Comment by wegank** [comment]
@ofborg build python3Packages.pyqt6

**Comment by wegank** [comment]
@ofborg build qt6.qtbase qt6.qtbase.passthru.tests

**Comment by wegank** [comment]
@NickCao I would like to know if you could merge the PR...

**Pull request merged**